### PR TITLE
Add support for complex union types

### PR DESCRIPTION
**Pull request by jasonxh**
A complex union will be converted to a struct type where field names are `member0`, `member1`, etc., in accordance with members of the union. This is consistent with the behavior when reading Parquet files. Field values are resolved following existing schema translation rules. The test case is a good example of this translation.

### CODE REVIEW

**Comment by sweb** [comment]
Hey!

are there any plans to merge this pull request in the next release? Since we are working with complex unions it would be of great help to us.

Thanks!
Florian

**Comment by jasonxh** [comment]
Rebased on top of current master. In the meantime, I'll be maintaining patched artifacts for 2.x and 3.x versions in this bintray repo:
https://bintray.com/jasonxh/maven/spark-avro

File: README.md [hunk 1]
```diff
@@ -101,12 +101,15 @@ This library supports reading all Avro types, with the exception of complex `uni
...
 
-In addition to the types listed above, it supports reading of three types of `union` types:
+In addition to the types listed above, it supports reading `union` types. The following three types are considered basic `union` types:
 
-1. `union(int, long)`
-2. `union(float, double)`
-3. `union(something, null)`, where `something` is one of the supported Avro types listed above or is one of the supported `union` types.
+1. `union(int, long)` will be mapped to `LongType`.
+2. `union(float, double)` will be mapped to `DoubleType`.
+3. `union(something, null)`, where `something` is any supported Avro type. This will be mapped to the same Spark SQL type as that of `something`, with `nullable` set to `true`.
```
Comment (JoshRosen) [inline README.md]: This makes me curious about the `union(union(int, long), null)` case. Let me quickly double-check to make sure that there's a test covering this corner-case to make sure that behavior hasn't changed here.

File: src/test/scala/com/databricks/spark/avro/AvroSuite.scala [hunk 2]
```diff
@@ -22,6 +22,8 @@ import java.nio.file.Files
 import java.sql.Timestamp
 import java.util.UUID
 
+import org.apache.avro.generic.GenericData.{EnumSymbol, Fixed}
```
Comment (JoshRosen) [inline src/test/scala/com/databricks/spark/avro/AvroSuite.scala]: This import should be grouped with the other Avro imports further down in this file.

File: README.md [hunk 3]
```diff
@@ -101,12 +101,15 @@ This library supports reading all Avro types, with the exception of complex `uni
...
+
+All other `union` types are considered complex. They will be mapped to `StructType` where field names are `member0`, `member1`, etc., in accordance with members of the `union`. This is consistent with the behavior when reading Parquet files.
```
Comment (JoshRosen) [inline README.md]: Thanks for remembering to add docs, by the way.

File: src/main/scala/com/databricks/spark/avro/SchemaConverters.scala [hunk 4]
```diff
@@ -264,12 +273,29 @@ object SchemaConverters {
...
-            case other => throw new IncompatibleSchemaException(
-              s"Cannot convert Avro schema to catalyst type because schema at path " +
-                s"${path.mkString(".")} is not compatible (avroType = $other, sqlType = $sqlType)" +
-                s" or this mix of union types is not supported (see README):\n" +
-                s"Source Avro schema: $sourceAvroSchema.\n" +
-                s"Target Catalyst type: $targetSqlType")
+            case other =>
+              sqlType match {
+                case t: StructType if t.fields.length == avroSchema.getTypes.size =>
+                  val fieldConverters = t.fields zip avroSchema.getTypes map {
```
Comment (JoshRosen) [inline src/main/scala/com/databricks/spark/avro/SchemaConverters.scala]: Minor style nit: we tend to prefer `.zip(..)` to infix notation.

File: src/main/scala/com/databricks/spark/avro/SchemaConverters.scala [hunk 5]
```diff
@@ -90,8 +90,17 @@ object SchemaConverters {
             SchemaType(LongType, nullable = false)
           case Seq(t1, t2) if Set(t1, t2) == Set(FLOAT, DOUBLE) =>
             SchemaType(DoubleType, nullable = false)
-          case other => throw new IncompatibleSchemaException(
-            s"This mix of union types is not supported (see README): $other")
+          case _ =>
+            // Convert complex unions to struct types where field names are member0, member1, etc.
+            // This is consistent with the behavior when reading Parquet files.
```
Comment (liancheng) [inline src/main/scala/com/databricks/spark/avro/SchemaConverters.scala]: By "reading Parquet files", I guess you mean reading Avro records as Parquet messages using parquet-avro?

File: README.md (hunk 1 above)
Comment (jasonxh) [inline README.md]: yep this should work as expected

File: src/main/scala/com/databricks/spark/avro/SchemaConverters.scala (hunk 4 above)
Comment (jasonxh) [inline src/main/scala/com/databricks/spark/avro/SchemaConverters.scala]: changed

File: src/test/scala/com/databricks/spark/avro/AvroSuite.scala (hunk 2 above)
Comment (jasonxh) [inline src/test/scala/com/databricks/spark/avro/AvroSuite.scala]: moved

File: src/main/scala/com/databricks/spark/avro/SchemaConverters.scala (hunk 5 above)
Comment (jasonxh) [inline src/main/scala/com/databricks/spark/avro/SchemaConverters.scala]: yes i meant avro parquet conversion. i've updated the comments

File: src/main/scala/com/databricks/spark/avro/SchemaConverters.scala [hunk 6]
```diff
@@ -264,12 +273,29 @@ object SchemaConverters {
...
+                    null
+                  } else {
+                    val i = GenericData.get().resolveUnion(avroSchema, item)
+                    val converted = new Array[Any](fieldConverters.length)
+                    converted(i) = fieldConverters(i)(item)
+                    new GenericRow(converted)
+                  }
+                case _ => throw new IncompatibleSchemaException(
+                  s"Cannot convert Avro schema to catalyst type because schema at path " +
+                    s"${path.mkString(".")} is not compatible " +
```
Comment (liancheng) [inline src/main/scala/com/databricks/spark/avro/SchemaConverters.scala]: Nit: I'd prefer "not supported" instead of "not compatible". "Compatible" is a term usually used in scenarios like schema evolution/merging, while here it means Spark SQL doesn't recognize some specific Avro schema.

**Comment by liancheng** [comment]
Also LGTM. Just two minor comments. Thanks!

File: src/main/scala/com/databricks/spark/avro/SchemaConverters.scala (hunk 6 above)
Comment (jasonxh) [inline src/main/scala/com/databricks/spark/avro/SchemaConverters.scala]: I was trying to keep the original wording (there's another reference right below). I believe it is referring to the incompatibility between source avro schema and target catalyst type. I'm fine with either wording. Let me know and I can change or leave both references.

File: README.md (hunk 1 above)
Comment (JoshRosen) [inline README.md]: Yep, I verified this as well.

File: src/main/scala/com/databricks/spark/avro/SchemaConverters.scala (hunk 6 above)
Comment (JoshRosen) [inline src/main/scala/com/databricks/spark/avro/SchemaConverters.scala]: I chatted with @liancheng about this and we think it's fine to keep the current wording.

**Comment by JoshRosen** [comment]
LGTM, so I'm going to merge this in to master now. Thanks for your contributions and patience, @jasonxh!
